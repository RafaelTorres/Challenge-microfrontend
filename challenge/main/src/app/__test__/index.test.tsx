import HomePage from '@root/src/pages';
import { act, fireEvent, render, RenderResult, screen } from '@testing-library/react';

// Mock dynamic component (Federation component load using lazy loading)
jest.mock('next/dynamic', () => () => {
  const DynamicComponent = ({ count = 0 }) => <p>{count}</p>;
  DynamicComponent.displayName = 'LoadableComponent';
  DynamicComponent.preload = jest.fn();
  return DynamicComponent;
});

const renderPage = async () => {
  let component = {} as RenderResult;

  await act(async () => {
    component = render(<HomePage />);
  });

  return component;
};

describe('Root Page', () => {
  beforeAll(() => {
    jest.clearAllMocks();
  });

  it('should render home page', async () => {
    const { container } = await renderPage();

    // Check if elements exist
    const title = screen.queryByText(/Counter - Module Federation/);
    expect(title).toBeInTheDocument();

    const buttonPlus = screen.queryByText('+');
    expect(buttonPlus).toBeInTheDocument();

    const buttonMinus = screen.queryByText('-');
    expect(buttonMinus).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should press on add item button and increase counter', async () => {
    const { container } = await renderPage();

    const button = screen.getByText('+');

    await act(() => {
      fireEvent.click(button);
    });

    const counterLabel = screen.queryByText('1');
    expect(counterLabel).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should press on add item button and decrease counter', async () => {
    const { container } = await renderPage();

    const button = screen.getByText('-');

    await act(() => {
      fireEvent.click(button);
    });

    const counterLabel = screen.queryByText('-1');
    expect(counterLabel).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
