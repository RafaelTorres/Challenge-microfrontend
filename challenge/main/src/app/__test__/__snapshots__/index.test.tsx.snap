// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root Page should press on add item button and decrease counter 1`] = `
<div>
  <main
    class="main"
  >
    <h2>
      Counter - Module Federation
    </h2>
    <p>
      -1
    </p>
    <div
      class="button-container"
    >
      <button
        class="button"
        type="button"
      >
        +
      </button>
      <button
        class="button"
        type="button"
      >
        -
      </button>
    </div>
  </main>
</div>
`;

exports[`Root Page should press on add item button and increase counter 1`] = `
<div>
  <main
    class="main"
  >
    <h2>
      Counter - Module Federation
    </h2>
    <p>
      1
    </p>
    <div
      class="button-container"
    >
      <button
        class="button"
        type="button"
      >
        +
      </button>
      <button
        class="button"
        type="button"
      >
        -
      </button>
    </div>
  </main>
</div>
`;

exports[`Root Page should render home page 1`] = `
<div>
  <main
    class="main"
  >
    <h2>
      Counter - Module Federation
    </h2>
    <p>
      0
    </p>
    <div
      class="button-container"
    >
      <button
        class="button"
        type="button"
      >
        +
      </button>
      <button
        class="button"
        type="button"
      >
        -
      </button>
    </div>
  </main>
</div>
`;
